File: examples_android/app/src/main/java/com/confirmit/testsdkapp/viewmodels/SurveyListViewModel.kt
package com.confirmit.testsdkapp.viewmodels

import android.util.Log
import com.confirmit.mobilesdk.SurveySDK
import com.confirmit.mobilesdk.database.externals.Server
import com.confirmit.mobilesdk.database.externals.Survey
import com.confirmit.testsdkapp.utils.ServerManager
import com.confirmit.testsdkapp.utils.ServerPref
import com.confirmit.testsdkapp.viewmodels.models.SurveyModel
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import java.io.Serializable

class SurveyListViewModel : Serializable {
    var surveys: MutableList<SurveyModel> = mutableListOf()

    init {
        try {
            SurveySDK.getSurveyList().forEach {
                surveys.add(SurveyModel(it))
            }
        } catch (_: Exception) {
        }
    }

    fun reloadSurveys() {
        try {
            surveys.clear()
            SurveySDK.getSurveyList().forEach {
                surveys.add(SurveyModel(it))
            }
        } catch (_: Exception) {
        }
    }

    fun downloadAllAsync(completion: (Boolean) -> Unit) = GlobalScope.async {
        val results = surveys.map {
            return@map Pair(it, SurveySDK.downloadSurveyAsync(it.survey.serverId, it.survey.surveyId))
        }

        val updatedSurveys = results.mapNotNull {
            try {
                val survey = it.first
                it.second.await()
                return@mapNotNull SurveySDK.getSurvey(survey.survey.serverId, survey.survey.surveyId)!!
            } catch (e: Exception) {
                return@mapNotNull null
            }
        }

        processSurveys(updatedSurveys, completion)
    }

    fun downloadAsync(serverId: String, surveyId: String, completion: (Boolean) -> Unit) = GlobalScope.async {
        try {
            val result = SurveySDK.downloadSurveyAsync(serverId, surveyId).await()
            if (result.result) {
                val updated = SurveySDK.getSurvey(serverId, surveyId)!!
                processSurveys(mutableListOf(updated), completion)
            } else {
                throw result.exception!!
            }

        } catch (e: Exception) {
            Log.e("MobileSDK", "Failed to downloadAsync survey", e)
            completion(false)
        }
    }

    fun setNewServer(serverId: String, clientId: String, clientSecret: String): Server {
        val server =  ServerManager.setNewConfiguration(serverId, clientId, clientSecret)
        ServerManager.writeServerToPref(serverId, ServerPref(clientId, clientSecret))
        return server
    }

    private fun processSurveys(updatedSurvey: List<Survey>, completion: (Boolean) -> Unit) {
        updatedSurvey.forEach { survey ->
            val existIndex = surveys.indexOfFirst { survey.serverId == it.survey.serverId && survey.surveyId == it.survey.surveyId }
            if (existIndex >= 0) {
                surveys[existIndex].update(survey)
            } else {
                try {
                    surveys.add(SurveyModel(survey))
                } catch (_: Exception) {
                }
            }
        }

        completion(updatedSurvey.isNotEmpty())
    }
}

File: examples_ios/MyApp/Sources/ViewModels/SurveyListViewModel.swift
import Foundation
import ConfirmitMobileSDK

typealias TableDataProcessCompletion = (_ hasUpdate: Bool) -> Void

class SurveyListViewModel {
    var surveys: [SurveyModel] = []
    var selectedSurvey: SurveyModel?
    
    init() {
        do {
            for survey in try SurveySDK.getSurveyList() {
                surveys.append(try SurveyModel(survey: survey))
            }
        } catch {}
    }
    
    func reloadSurveys() {
        do {
            var exists = [String]()
            let curSurveys = try SurveySDK.getSurveyList()
            for survey in curSurveys {
                exists.append("\(survey.serverId)_\(survey.surveyId)")
            }
            
            surveys.removeAll {
                let key = "\($0.survey.serverId)_\($0.survey.surveyId)"
                return !exists.contains(key)
            }
            
            processSurveys(updatedSurvey: curSurveys) { _ in }
        } catch {}
    }
    
    func downloadAll(completion: @escaping TableDataProcessCompletion) {
        let dispatchGroup = DispatchGroup()
        
        var updatedSurveys = [Survey]()
        for survey in surveys {
            dispatchGroup.enter()
            let serverId = survey.survey.serverId
            let surveyId = survey.survey.surveyId
            SurveySDK.downloadSurvey(serverId: serverId, surveyId: surveyId) { result, _ in
                if result {
                    do {
                        if let updated = try SurveySDK.getSurvey(serverId: serverId, surveyId: surveyId) {
                            updatedSurveys.append(updated)
                        }
                    } catch {
                        // Log
                    }
                }
                dispatchGroup.leave()
            }
        }
        
        dispatchGroup.notify(queue: .main, work: DispatchWorkItem(block: {
            self.processSurveys(updatedSurvey: updatedSurveys, completion: completion)
        }))
    }
    
    func download(serverId: String, surveyId: String, completion: @escaping TableDataProcessCompletion) {
        SurveySDK.downloadSurvey(serverId: serverId, surveyId: surveyId) { result, _ in
            if result {
                do {
                    if let updated = try SurveySDK.getSurvey(serverId: serverId, surveyId: surveyId) {
                        self.processSurveys(updatedSurvey: [updated], completion: completion)
                    }
                } catch {
                    // Log
                }
            }
        }
    }
    
    private func processSurveys(updatedSurvey: [Survey], completion: @escaping TableDataProcessCompletion) {
        for survey in updatedSurvey {
            if let existIndex = self.surveys.firstIndex(where: { (model) -> Bool in
                survey.serverId == model.survey.serverId && survey.surveyId == model.survey.surveyId
            }) {
                surveys[existIndex].update(survey: survey)
            } else {
                do {
                    surveys.append(try SurveyModel(survey: survey))
                } catch {}
            }
        }
        
        completion(updatedSurvey.count > 0)
    }
}
